File: 0451-sort-characters-by-frequency/0451-sort-characters-by-frequency.cpp
class Solution {
public:
    string frequencySort(string s) {
        unordered_map<char,int> hashMap; 
        for(char& c : s) hashMap[c]++;

        sort(s.begin(), s.end(), [&](char& a, char& b){
            if(hashMap[a] == hashMap[b]) return a < b;
            return hashMap[a] > hashMap[b]; 
        });

        return s; 
    }
};

File: 0451-sort-characters-by-frequency/0451-sort-characters-by-frequency.java
class Solution {
    public String frequencySort(String s) {
        Map<Character,Integer> hashMap = new HashMap<>(); 
        for(char c : s.toCharArray()) hashMap.put(c, hashMap.getOrDefault(c,0) + 1); 
        List<Character> charList = new ArrayList<>(); 
        for(char c : s.toCharArray()) charList.add(c); 
        Collections.sort(charList, (a,b) -> {
            int freqA = hashMap.get(a);
            int freqB = hashMap.get(b); 
            if(freqB == freqA){
                return b - a;  
            }
            return freqB - freqA; 
        });

        StringBuilder sb = new StringBuilder(); 
        for(char c : charList){
            sb.append(c); 
        }


        return sb.toString(); 
    }
}